internal/pkg/api/handleCheckin.go

```diff
@@ -445,23 +439,27 @@ func (ct *CheckinT) verifyActionExists(vCtx context.Context, vSpan *apm.Span, ag
 // if the agent doc and checkin details are both nil the method is a nop
 // if the checkin upgrade_details is nil but there was a previous value in the agent doc, fleet-server treats it as a successful upgrade
 // otherwise the details are validated; action_id is checked and upgrade_details.metadata is validated based on upgrade_details.state and the agent doc is updated.
-func (ct *CheckinT) processUpgradeDetails(ctx context.Context, agent *model.Agent, details *UpgradeDetails) error {
+func (ct *CheckinT) processUpgradeDetails(ctx context.Context, agent *model.Agent, details *UpgradeDetails, checkinOpts []checkin.Option) ([]checkin.Option, error) {
 	if details == nil {
-		err := ct.markUpgradeComplete(ctx, agent)
-		if err != nil {
-			return err
+		if agent.UpgradeDetails == nil {
+			return checkinOpts, nil
 		}
-		return nil
+		checkinOpts = append(checkinOpts,
+			checkin.WithUpgradeDetails(nil),
+			checkin.WithUpgradeStartedAt(nil),
+			checkin.WithUpgradeStatus(nil),
+		)
```

Comment:
I think as @pchila mentioned, we need to set `dl.FieldUpgradedAt` here to mark a successful upgrade (iirc @ycombinator did the implementation and may know more, or at least where the state-diagram for this is)